zebrad/src/components/tracing.rs

```diff
@@ -125,3 +134,74 @@ To set the filter, POST the new filter string to /filter:
     };
     Ok(rsp)
 }
+
+#[derive(Clone)]
+pub(crate) struct FlameGrapher {
+    guard: Arc<tracing_flame::FlushGuard<BufWriter<File>>>,
+    path: PathBuf,
+}
+
+impl FlameGrapher {
+    fn make_flamegraph(&self) -> Result<(), Report> {
+        self.guard.flush()?;
+        let out_path = self.path.with_extension("svg");
+        let inf = File::open(&self.path)?;
+        let reader = BufReader::new(inf);
+
+        let out = File::create(out_path)?;
+        let writer = BufWriter::new(out);
+
+        let mut opts = inferno::flamegraph::Options::default();
+        info!("writing flamegraph to disk...");
+        inferno::flamegraph::from_reader(&mut opts, reader, writer)?;
+
+        Ok(())
+    }
+}
+
+impl Drop for FlameGrapher {
+    fn drop(&mut self) {
+        match self.make_flamegraph() {
+            Ok(()) => {}
+            Err(report) => {
+                warn!(
+                    "Error while constructing flamegraph during shutdown: {:?}",
+                    report
+                );
+            }
+        }
+    }
+}
+
+pub(crate) fn init(level: EnvFilter) -> (Tracing, Option<FlameGrapher>) {
+    // Construct a tracing subscriber with the supplied filter and enable reloading.
+    let builder = tracing_subscriber::FmtSubscriber::builder()
+        .with_env_filter(level)
+        .with_filter_reloading();
+    let filter_handle = builder.reload_handle();
+    let subscriber = builder.finish().with(tracing_error::ErrorLayer::default());
+
+    let guard = if let Ok(flamegraph_path) = std::env::var(FLAMEGRAPH_ENV) {
+        let flamegraph_path = Path::new(&flamegraph_path).with_extension("folded");
+        let (flame_layer, guard) = tracing_flame::FlameLayer::with_file(&flamegraph_path).unwrap();
+        let flame_layer = flame_layer
+            .with_empty_samples(false)
+            .with_threads_collapsed(true);
+        subscriber.with(flame_layer).init();
+        Some(FlameGrapher {
+            guard: Arc::new(guard),
+            path: flamegraph_path,
+        })
+    } else {
+        subscriber.init();
+        None
+    };
+
+    (filter_handle.into(), guard)
```

Comment:
Rather than changing Abscissa to have `impl From` on a filter handle, what if we created a second `Tracing` constructor that took a pre-constructed subscriber?

Reply:
I don't think that can be done, there's no generic way to get access to the Filter handle from a subscriber, and the `Tracing` type is defined in such a way where it requires a filter handle as one of its data members.

Reply:
Okay, but we can change the definition of the `Tracing` type or any of its constructors.  What if we had it take a pre-constructed subscriber *and* a filter handle?

Reply:
What does this accomplish? We'd still have to make this change upstream in abscissa. I'm not opposed to this I just don't understand the motivation.

Reply:
The motivation is to be able to fully control the subscriber, set up a registry, etc.  But I guess this isn't necessary, since the `Tracing` component can just have `Tracing::new(...)` that does some default configuration and calls `set_global_default` and a `Tracing::from_handle()` (your current `TryFrom` impl) that offloads the setup to the caller. 

I think it was a mistake to upstream our tracing handling; we assumed that we wouldn't need to do detailed configuration of tracing behavior, and that's turned out not to be the case.  Rather than try to jump through hoops to make this work, we should instead take control of our tracing component and configure it properly.  We can do this in a later PR.